File: windows-app/Micropad.App/ViewModels/MacrosViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Microsoft.Win32;
using Micropad.Core;
using Micropad.Core.Models;
using Micropad.Services.Input;

namespace Micropad.App.ViewModels;

public partial class MacrosViewModel : ObservableObject
{
    private readonly MacroRecorder _recorder;
    private readonly Micropad.Services.Storage.LocalMacroStorage _macroStorage;

    [ObservableProperty]
    private ObservableCollection<MacroStep> _steps = new();

    [ObservableProperty]
    private string _macroName = "New Macro";

    [ObservableProperty]
    private string _statusText = "Create a macro or record one.";

    [ObservableProperty]
    private bool _isRecording;

    [ObservableProperty]
    private ObservableCollection<MicroSlot> _slots = new();

    [ObservableProperty]
    private int _selectedSlotIndex = -1;

    [ObservableProperty]
    private string _selectedSequence = string.Empty;

    [ObservableProperty]
    private string _urlInput = string.Empty;

    [ObservableProperty]
    private string _applicationPathInput = string.Empty;

    [ObservableProperty]
    private string _selectedBrowser = "Default";

    [ObservableProperty]
    private string _textToAdd = string.Empty;

    [ObservableProperty]
    private MacroStep? _selectedStep;

    public bool IsInspectorVisible => SelectedStep != null;
    public bool IsDelayStep => SelectedStep?.Action == "delay";
    public bool IsTextStep => SelectedStep?.Action == "textType";
    public bool IsKeyStep => SelectedStep != null && (SelectedStep.Action == "keyPress" || SelectedStep.Action == "keyDown" || SelectedStep.Action == "keyUp");

    partial void OnSelectedStepChanged(MacroStep? value)
    {
        OnPropertyChanged(nameof(IsInspectorVisible));
        OnPropertyChanged(nameof(IsDelayStep));
        OnPropertyChanged(nameof(IsTextStep));
        OnPropertyChanged(nameof(IsKeyStep));
    }

    public ObservableCollection<string> Browsers { get; } = new()
    {
        "Default",
        "Chrome",
        "Edge",
        "Firefox",
        "Opera",
        "Brave"
    };

    public ObservableCollection<MacroTag> AllTags { get; } = new();
    public IEnumerable<IGrouping<string, MacroTag>> TagGroups => AllTags.GroupBy(t => t.Group).OrderBy(g => g.Key);

    public bool IsNotRecording => !IsRecording;

    public int StepCount => Steps.Count;

    public bool IsStepsEmpty => Steps.Count == 0;

    public MicroSlot? SelectedSlot => SelectedSlotIndex >= 0 && SelectedSlotIndex < Slots.Count ? Slots[SelectedSlotIndex] : null;

    partial void OnIsRecordingChanged(bool value) => OnPropertyChanged(nameof(IsNotRecording));
    partial void OnSelectedSlotIndexChanged(int value) => UpdateSelectedSequence();
    partial void OnSelectedSequenceChanged(string value)
    {
        if (SelectedSlot != null)
        {
            SelectedSlot.Sequence = value;
            OnPropertyChanged(nameof(Slots));
        }
    }

    public MacrosViewModel(MacroRecorder recorder, Micropad.Services.Storage.LocalMacroStorage macroStorage)
    {
        _recorder = recorder;
        _macroStorage = macroStorage;
        InitSlots();
        foreach (var t in MacroTagCatalog.GetAll())
            AllTags.Add(t);
    }

    private void InitSlots()
    {
        Slots.Clear();
        for (int i = 0; i < 12; i++)
            Slots.Add(new MicroSlot { Index = i, Label = $"K{i + 1}", IsEncoder = false });
        Slots.Add(new MicroSlot { Index = 12, Label = "E1", IsEncoder = true });
        Slots.Add(new MicroSlot { Index = 13, Label = "E2", IsEncoder = true });
    }

    private void UpdateSelectedSequence()
    {
        SelectedSequence = SelectedSlot?.Sequence ?? string.Empty;
    }

    [RelayCommand]
    private void SelectSlot(int index)
    {
        if (index >= 0 && index < Slots.Count)
        {
            SelectedSlotIndex = index;
            StatusText = $"Editing {Slots[index].Label}. Drag tags here or type below.";
        }
    }

    [RelayCommand]
    private void AppendTag(string? tag)
    {
        if (string.IsNullOrEmpty(tag)) return;
        if (SelectedSlot != null)
        {
            SelectedSlot.Sequence += tag;
            SelectedSequence = SelectedSlot.Sequence;
            OnPropertyChanged(nameof(Slots));
        }
        else
        {
            StatusText = "Select a key or encoder slot first.";
        }
    }

    public void DropTagOnSlot(int slotIndex, string tag)
    {
        if (slotIndex < 0 || slotIndex >= Slots.Count || string.IsNullOrEmpty(tag)) return;
        Slots[slotIndex].Sequence += tag;
        SelectedSlotIndex = slotIndex;
        SelectedSequence = Slots[slotIndex].Sequence;
        OnPropertyChanged(nameof(Slots));
        StatusText = $"Dropped {tag} on {Slots[slotIndex].Label}.";
    }

    [RelayCommand]
    private void ClearSlot(int index)
    {
        if (index >= 0 && index < Slots.Count)
        {
            Slots[index].Sequence = string.Empty;
            if (SelectedSlotIndex == index) SelectedSequence = string.Empty;
            OnPropertyChanged(nameof(Slots));
            StatusText = $"Cleared {Slots[index].Label}.";
        }
    }

    [RelayCommand]
    private void StartRecording()
    {
        _recorder.StartRecording();
        IsRecording = true;
        Steps.Clear();
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Recording... Press keys. Click Stop when done.";
    }

    [RelayCommand]
    private void StopRecording()
    {
        var recorded = _recorder.StopRecording();
        IsRecording = false;
        Steps.Clear();
        foreach (var s in recorded)
        {
            Steps.Add(s);
        }
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = $"Recorded {Steps.Count} step(s).";
    }

    [RelayCommand]
    private void AddDelay()
    {
        Steps.Add(new MacroStep { Action = "delay", DelayMs = 100 });
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Added delay step.";
    }

    [RelayCommand]
    private void AddKeyPress(string keyName)
    {
        if (string.IsNullOrEmpty(keyName)) keyName = "A";
        Steps.Add(new MacroStep { Action = "keyPress", Key = keyName });
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
    }

    [RelayCommand]
    private void AddText()
    {
        var text = (TextToAdd ?? "").Trim();
        if (string.IsNullOrEmpty(text))
        {
            StatusText = "Enter text first.";
            return;
        }
        Steps.Add(new MacroStep { Action = "textType", Text = text });
        TextToAdd = string.Empty;
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Added text step.";
    }

    [RelayCommand]
    private void ClearMacro()
    {
        Steps.Clear();
        SelectedStep = null;
        MacroName = "New Macro";
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Macro cleared. Record or add steps to start fresh.";
    }

    [RelayCommand]
    private void RemoveStep(MacroStep? step)
    {
        if (step != null && Steps.Contains(step))
        {
            if (step == SelectedStep) SelectedStep = null;
            Steps.Remove(step);
            OnPropertyChanged(nameof(StepCount));
            OnPropertyChanged(nameof(IsStepsEmpty));
        }
    }

    [RelayCommand]
    private void MoveStepUp(MacroStep? step)
    {
        if (step == null || !Steps.Contains(step)) return;
        var i = Steps.IndexOf(step);
        if (i <= 0) return;
        Steps.Move(i, i - 1);
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Step moved up.";
    }

    [RelayCommand]
    private void MoveStepDown(MacroStep? step)
    {
        if (step == null || !Steps.Contains(step)) return;
        var i = Steps.IndexOf(step);
        if (i < 0 || i >= Steps.Count - 1) return;
        Steps.Move(i, i + 1);
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Step moved down.";
    }

    [RelayCommand]
    private void ClearSteps()
    {
        Steps.Clear();
        OnPropertyChanged(nameof(StepCount));
        OnPropertyChanged(nameof(IsStepsEmpty));
        StatusText = "Steps cleared.";
    }

    [RelayCommand]
    private void SaveMacro()
    {
        var name = (MacroName ?? "").Trim();
        if (string.IsNullOrEmpty(name))
        {
            StatusText = "Enter a macro name first.";
            return;
        }
        if (Steps.Count == 0)
        {
            StatusText = "Add steps or record a macro first.";
            return;
        }
        _macroStorage.SaveMacro(name, GetStepsCopy());
        OnPropertyChanged(nameof(StepCount));
        StatusText = $"Saved macro '{name}'.";
    }

    [RelayCommand]
    private void ExportMacro()
    {
        var name = (MacroName ?? "").Trim();
        if (string.IsNullOrEmpty(name)) name = "Unnamed";
        if (Steps.Count == 0)
        {
            StatusText = "Add steps or record a macro first.";
            return;
        }
        var dialog = new SaveFileDialog
        {
            Filter = "Macro JSON (*.json)|*.json|All files (*.*)|*.*",
            DefaultExt = "json",
            FileName = name + ".json"
        };
        if (dialog.ShowDialog() != true) return;
        try
        {
            var asset = new MacroAsset
            {
                Name = name,
                Steps = GetStepsCopy(),
                MacroId = Guid.NewGuid().ToString("N"),
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };
            _macroStorage.ExportMacroAsset(asset, dialog.FileName);
            StatusText = $"Exported to {dialog.FileName}";
        }
        catch (Exception ex)
        {
            StatusText = $"Export failed: {ex.Message}";
        }
    }

    [RelayCommand]
    private void ImportMacro()
    {
        var dialog = new OpenFileDialog
        {
            Filter = "Macro JSON (*.json)|*.json|All files (*.*)|*.*"
        };
        if (dialog.ShowDialog() != true) return;
        try
        {
            var asset = _macroStorage.ImportMacroAsset(dialog.FileName);
            if (asset != null)
            {
                MacroName = asset.Name;
                SelectedStep = null;
                Steps.Clear();
                foreach (var s in asset.Steps)
                    Steps.Add(s);
                OnPropertyChanged(nameof(StepCount));
                OnPropertyChanged(nameof(IsStepsEmpty));
                StatusText = $"Imported '{asset.Name}'. Assign it from Profiles.";
            }
            else
                StatusText = "Import failed or file invalid.";
        }
        catch (Exception ex)
        {
            StatusText = $"Import failed: {ex.Message}";
        }
    }

    public List<MacroStep> GetStepsCopy()
    {
        return Steps.Select(s => new MacroStep
        {
            Action = s.Action,
            Key = s.Key,
            DelayMs = s.DelayMs,
            Text = s.Text,
            Value = s.Value,
            VkCode = s.VkCode,
            MediaFunction = s.MediaFunction
        }).ToList();
    }

    [RelayCommand]
    private void InsertUrl()
    {
        if (string.IsNullOrWhiteSpace(UrlInput))
        {
            StatusText = "Enter a URL first.";
            return;
        }
        if (SelectedSlot == null)
        {
            StatusText = "Select a key or encoder slot first.";
            return;
        }

        var url = UrlInput.Trim();
        if (!url.StartsWith("http://", StringComparison.OrdinalIgnoreCase) && 
            !url.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
        {
            url = "https://" + url;
        }

        string browserTag;
        switch (SelectedBrowser)
        {
            case "Chrome":
                browserTag = "{RUN:chrome.exe " + url + "}";
                break;
            case "Edge":
                browserTag = "{RUN:msedge.exe " + url + "}";
                break;
            case "Firefox":
                browserTag = "{RUN:firefox.exe " + url + "}";
                break;
            case "Opera":
                browserTag = "{RUN:opera.exe " + url + "}";
                break;
            case "Brave":
                browserTag = "{RUN:brave.exe " + url + "}";
                break;
            default:
                // Use Windows shell start command for default browser
                browserTag = "{RUN:start " + url + "}";
                break;
        }

        SelectedSlot.Sequence += browserTag;
        SelectedSequence = SelectedSlot.Sequence;
        OnPropertyChanged(nameof(Slots));
        StatusText = $"Added URL to {SelectedSlot.Label}.";
        UrlInput = string.Empty;
    }

    [RelayCommand]
    private void InsertApplication()
    {
        if (string.IsNullOrWhiteSpace(ApplicationPathInput))
        {
            StatusText = "Enter an application path first.";
            return;
        }
        if (SelectedSlot == null)
        {
            StatusText = "Select a key or encoder slot first.";
            return;
        }

        var path = ApplicationPathInput.Trim();
        var appTag = "{RUN:" + path + "}";
        
        SelectedSlot.Sequence += appTag;
        SelectedSequence = SelectedSlot.Sequence;
        OnPropertyChanged(nameof(Slots));
        StatusText = $"Added application to {SelectedSlot.Label}.";
        ApplicationPathInput = string.Empty;
    }

    [RelayCommand]
    private void BrowseApplication()
    {
        var dialog = new OpenFileDialog
        {
            Filter = "Executable files (*.exe)|*.exe|All files (*.*)|*.*",
            Title = "Select Application"
        };

        if (dialog.ShowDialog() == true)
        {
            ApplicationPathInput = dialog.FileName;
        }
    }
}

File: macos-app/Micropad/ViewModels/MacrosViewModel.swift
//
//  MacrosViewModel.swift
//  Micropad
//
//  ViewModel for macro/shortcut builder
//

import Foundation
import Combine

@MainActor
class MacrosViewModel: ObservableObject {
    @Published var slots: [MicroSlot] = []
    @Published var selectedSlotIndex: Int? = nil
    @Published var selectedSequence: String = ""
    @Published var allTags: [MacroTag] = []
    @Published var urlInput: String = ""
    @Published var applicationPathInput: String = ""
    @Published var selectedBrowser: String = "Default"
    @Published var selectedProfile: Profile? = nil
    @Published var availableProfiles: [Profile] = []
    
    let browsers = ["Default", "Chrome", "Safari", "Firefox", "Edge", "Opera", "Brave"]
    
    
    var selectedSlot: MicroSlot? {
        guard let index = selectedSlotIndex, index >= 0, index < slots.count else { return nil }
        return slots[index]
    }
    
    init() {
        initSlots()
        allTags = MacroTagCatalog.getAll()
        loadProfiles()
    }
    
    private func loadProfiles() {
        availableProfiles = StorageService.shared.loadProfiles()
        if availableProfiles.isEmpty {
            // Create default profile
            createDefaultProfile()
        } else {
            selectedProfile = availableProfiles.first
            loadProfileMacros()
        }
    }
    
    private func createDefaultProfile() {
        let newProfile = Profile(id: 0, name: "Profile 1", version: 1)
        availableProfiles = [newProfile]
        selectedProfile = newProfile
        StorageService.shared.saveProfiles(availableProfiles)
    }
    
    func selectProfile(_ profile: Profile) {
        saveCurrentProfileMacros()
        selectedProfile = profile
        loadProfileMacros()
    }
    
    func createNewProfile(name: String, id: Int) {
        saveCurrentProfileMacros()
        
        // Check if ID already exists
        if availableProfiles.contains(where: { $0.id == id }) {
            return
        }
        
        let newProfile = Profile(id: id, name: name, version: 1)
        availableProfiles.append(newProfile)
        selectedProfile = newProfile
        StorageService.shared.saveProfiles(availableProfiles)
        loadProfileMacros()
    }
    
    func updateProfileName(_ name: String) {
        guard var profile = selectedProfile else { return }
        let trimmedName = name.trimmingCharacters(in: .whitespaces)
        guard !trimmedName.isEmpty else { return }
        profile.name = trimmedName
        if let index = availableProfiles.firstIndex(where: { $0.id == profile.id }) {
            availableProfiles[index] = profile
            selectedProfile = profile
            StorageService.shared.saveProfiles(availableProfiles)
        }
    }
    
    func refreshProfiles() {
        let currentProfileId = selectedProfile?.id
        availableProfiles = StorageService.shared.loadProfiles()
        if let id = currentProfileId, let profile = availableProfiles.first(where: { $0.id == id }) {
            selectedProfile = profile
            loadProfileMacros()
        } else if let first = availableProfiles.first {
            selectedProfile = first
            loadProfileMacros()
        }
    }
    
    private func loadProfileMacros() {
        guard let profile = selectedProfile else { return }
        
        // Load macro sequences from profile keys
        for i in 0..<12 {
            if i < profile.keys.count {
                let keyConfig = profile.keys[i]
                if let sequence = keyConfig.macroSequence {
                    slots[i].sequence = sequence
                } else {
                    slots[i].sequence = ""
                }
            } else {
                slots[i].sequence = ""
            }
        }
        
        // Load encoder sequences if any
        // Note: Encoders don't have macro sequences in the current model, but we can add them
    }
    
    func saveCurrentProfileMacros() {
        guard var profile = selectedProfile else { return }
        
        // Save macro sequences to profile keys
        for i in 0..<12 {
            if i < slots.count {
                let sequence = slots[i].sequence
                if i < profile.keys.count {
                    profile.keys[i].macroSequence = sequence.isEmpty ? nil : sequence
                    if !sequence.isEmpty {
                        profile.keys[i].type = .macro
                    }
                } else {
                    // Create new KeyConfig with macro sequence
                    var newKey = KeyConfig(
                        index: i,
                        type: sequence.isEmpty ? .none : .macro,
                        modifiers: 0,
                        key: 0,
                        function: 0,
                        action: 0,
                        value: 0,
                        profileId: profile.id
                    )
                    newKey.macroSequence = sequence.isEmpty ? nil : sequence
                    profile.keys.append(newKey)
                }
            }
        }
        
        // Update profile in list
        if let index = availableProfiles.firstIndex(where: { $0.id == profile.id }) {
            availableProfiles[index] = profile
            selectedProfile = profile
            StorageService.shared.saveProfiles(availableProfiles)
        }
    }
    
    private func initSlots() {
        slots = []
        for i in 0..<12 {
            slots.append(MicroSlot(index: i, label: "K\(i + 1)", isEncoder: false))
        }
        slots.append(MicroSlot(index: 12, label: "E1", isEncoder: true))
        slots.append(MicroSlot(index: 13, label: "E2", isEncoder: true))
    }
    
    func selectSlot(_ index: Int) {
        selectedSlotIndex = index
        updateSelectedSequence()
    }
    
    private func updateSelectedSequence() {
        selectedSequence = selectedSlot?.sequence ?? ""
    }
    
    func updateSelectedSequence(_ newSequence: String) {
        guard let slot = selectedSlot else { return }
        slot.sequence = newSequence
        selectedSequence = newSequence
        saveCurrentProfileMacros()
    }
    
    func appendTag(_ tag: String) {
        guard let slot = selectedSlot else { return }
        slot.sequence += tag
        selectedSequence = slot.sequence
        saveCurrentProfileMacros()
    }
    
    func dropTagOnSlot(_ slotIndex: Int, tag: String) {
        guard slotIndex >= 0, slotIndex < slots.count else { return }
        slots[slotIndex].sequence += tag
        selectedSlotIndex = slotIndex
        selectedSequence = slots[slotIndex].sequence
        saveCurrentProfileMacros()
    }
    
    func insertUrl() {
        guard let slot = selectedSlot else { return }
        var url = urlInput.trimmingCharacters(in: .whitespaces)
        if !url.hasPrefix("http://") && !url.hasPrefix("https://") {
            url = "https://" + url
        }
        
        let browserTag: String
        switch selectedBrowser {
        case "Chrome":
            browserTag = "{RUN:open -a \"Google Chrome\" \(url)}"
        case "Safari":
            browserTag = "{RUN:open -a Safari \(url)}"
        case "Firefox":
            browserTag = "{RUN:open -a Firefox \(url)}"
        case "Edge":
            browserTag = "{RUN:open -a \"Microsoft Edge\" \(url)}"
        case "Opera":
            browserTag = "{RUN:open -a Opera \(url)}"
        case "Brave":
            browserTag = "{RUN:open -a Brave \(url)}"
        default:
            browserTag = "{RUN:open \(url)}"
        }
        
        slot.sequence += browserTag
        selectedSequence = slot.sequence
        urlInput = ""
        saveCurrentProfileMacros()
    }
    
    func insertApplication() {
        guard let slot = selectedSlot else { return }
        let path = applicationPathInput.trimmingCharacters(in: .whitespaces)
        let appTag = "{RUN:open -a \"\(path)\"}"
        slot.sequence += appTag
        selectedSequence = slot.sequence
        applicationPathInput = ""
        saveCurrentProfileMacros()
    }
    
    func clearSlot(_ index: Int) {
        guard index >= 0, index < slots.count else { return }
        slots[index].sequence = ""
        if selectedSlotIndex == index {
            selectedSequence = ""
        }
        saveCurrentProfileMacros()
    }
    
    deinit {
        saveCurrentProfileMacros()
    }
}
